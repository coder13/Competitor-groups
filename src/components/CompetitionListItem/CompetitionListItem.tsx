import classNames from 'classnames';
import { hasFlag } from 'country-flag-icons';
import getUnicodeFlagIcon from 'country-flag-icons/unicode';
import { Link } from 'react-router-dom';
import { formatDateRange } from '@/lib/time';

interface CompetitionListItemProps {
  id: string;
  name: string;
  start_date: string;
  end_date: string;
  country_iso2: string;
  city: string;
  isLive: boolean;
  isBookmarked?: boolean;
}

export const CompetitionListItem = ({
  id,
  name,
  start_date,
  end_date,
  country_iso2,
  city,
  isLive,
  isBookmarked,
}: CompetitionListItemProps) => {
  const endDate = new Date(
    new Date(end_date).getTime() + 1000 * 60 * new Date().getTimezoneOffset(),
  );

  return (
    <Link to={`/competitions/${id}`}>
      <li
        className={classNames(
          'border bg-white list-none rounded-md px-2 py-2 flex cursor-pointer hover:bg-slate-100 group transition-colors my-1 flex-row',
          {
            'opacity-50': endDate < new Date(Date.now() - 1000 * 60 * 60 * 24),
          },
        )}>
        {hasFlag(country_iso2) && (
          <div className="flex flex-shrink mr-2 text-2xl"> {getUnicodeFlagIcon(country_iso2)} </div>
        )}
        <div className="flex-1">
          <p className="font-normal leading-1"> {name} </p>{' '}
          <p className="text-gray-600 text-sm leading-1">
            {formatDateRange(start_date, end_date)}
            {'  '}&#8211;{'  '}
            {city}
          </p>
        </div>
        {isLive && (
          <div className="flex flex-shrink text-2xl items-center px-2">
            <span className="fa fa-tower-broadcast text-green-500" />
          </div>
        )}
        {isBookmarked && !isLive && (
          <div className="flex flex-shrink text-2xl items-center px-2">
            <span className="fa fa-bookmark text-yellow-500" />
          </div>
        )}
      </li>
    </Link>
  );
};
