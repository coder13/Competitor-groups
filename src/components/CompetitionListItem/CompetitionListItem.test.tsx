import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { CompetitionListItem } from './CompetitionListItem';

const baseProps = {
  id: '123',
  name: 'Test Competition',
  start_date: '2024-01-01',
  end_date: '2024-01-02',
  country_iso2: 'US',
  city: 'New York',
  isLive: false,
  isBookmarked: false,
};

function renderWithRouter(ui: React.ReactElement) {
  return render(ui, {
    wrapper: MemoryRouter,
  });
}

describe('CompetitionListItem', () => {
  it('renders with all props (live and bookmarked)', () => {
    const { baseElement } = renderWithRouter(
      <CompetitionListItem {...baseProps} isLive={true} isBookmarked={true} />,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('renders with only isLive true', () => {
    renderWithRouter(<CompetitionListItem {...baseProps} isLive={true} isBookmarked={false} />);
    expect(screen.getByRole('listitem').querySelector('.fa-tower-broadcast')).toBeInTheDocument();
    expect(screen.getByRole('listitem').querySelector('.fa-bookmark')).not.toBeInTheDocument();
  });

  it('renders with only isBookmarked true', () => {
    renderWithRouter(<CompetitionListItem {...baseProps} isLive={false} isBookmarked={true} />);
    expect(screen.getByRole('listitem').querySelector('.fa-bookmark')).toBeInTheDocument();
    expect(
      screen.getByRole('listitem').querySelector('.fa-tower-broadcast'),
    ).not.toBeInTheDocument();
  });

  it('renders with neither isLive nor isBookmarked', () => {
    renderWithRouter(<CompetitionListItem {...baseProps} isLive={false} isBookmarked={false} />);
    expect(
      screen.getByRole('listitem').querySelector('.fa-tower-broadcast'),
    ).not.toBeInTheDocument();
    expect(screen.getByRole('listitem').querySelector('.fa-bookmark')).not.toBeInTheDocument();
  });

  it('shows flag if country_iso2 is valid', () => {
    renderWithRouter(<CompetitionListItem {...baseProps} />);
    expect(screen.getByText('ðŸ‡ºðŸ‡¸')).toBeInTheDocument();
  });

  it('applies opacity-50 class if competition is in the past', () => {
    const pastProps = {
      ...baseProps,
      end_date: '2000-01-01',
    };
    renderWithRouter(<CompetitionListItem {...pastProps} />);
    expect(screen.getByRole('listitem').className).toMatch(/opacity-50/);
  });

  it('matches snapshot (default)', () => {
    const { baseElement } = renderWithRouter(<CompetitionListItem {...baseProps} />);
    expect(baseElement).toMatchSnapshot();
  });
});
