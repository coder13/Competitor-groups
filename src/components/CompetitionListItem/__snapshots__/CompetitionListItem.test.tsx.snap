// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompetitionListItem matches snapshot (default) 1`] = `
<body>
  <div>
    <a
      href="/competitions/123"
    >
      <li
        class="border bg-white list-none rounded-md px-2 py-2 flex cursor-pointer hover:bg-slate-100 group transition-colors my-1 flex-row opacity-50"
      >
        <div
          class="flex flex-shrink mr-2 text-2xl"
        >
           
          ðŸ‡ºðŸ‡¸
           
        </div>
        <div
          class="flex-1"
        >
          <p
            class="font-normal leading-1"
          >
             
            Test Competition
             
          </p>
           
          <p
            class="text-gray-600 text-sm leading-1"
          >
            Jan 1â€‰â€“â€‰2, 2024
              
            â€“
              
            New York
          </p>
        </div>
      </li>
    </a>
  </div>
</body>
`;

exports[`CompetitionListItem renders with all props (live and bookmarked) 1`] = `
<body>
  <div>
    <a
      href="/competitions/123"
    >
      <li
        class="border bg-white list-none rounded-md px-2 py-2 flex cursor-pointer hover:bg-slate-100 group transition-colors my-1 flex-row opacity-50"
      >
        <div
          class="flex flex-shrink mr-2 text-2xl"
        >
           
          ðŸ‡ºðŸ‡¸
           
        </div>
        <div
          class="flex-1"
        >
          <p
            class="font-normal leading-1"
          >
             
            Test Competition
             
          </p>
           
          <p
            class="text-gray-600 text-sm leading-1"
          >
            Jan 1â€‰â€“â€‰2, 2024
              
            â€“
              
            New York
          </p>
        </div>
        <div
          class="flex flex-shrink text-2xl items-center px-2"
        >
          <span
            class="fa fa-tower-broadcast text-green-500"
          />
        </div>
      </li>
    </a>
  </div>
</body>
`;
